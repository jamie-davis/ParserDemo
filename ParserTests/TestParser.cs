using Xunit;
using Parser;
using TestConsoleLib;
using TestConsoleLib.Testing;
using TestConsole.OutputFormatting;
using System.Linq;

namespace ParserTests
{
    public class TestParser
    {
        [Fact]
        public void InputIsConverted()
        {
            //Arrange
            var testStrings = new [] {
                "1",
                "32-5",
                "(100 * 3)",
                "100 + (3 * 5)",
                "(3 * 5) + 100",
                "1 + 2 + 3",
                "1 + (3 * (5 + 6))",
                "1 +",
                "(1 + 3",
                "(",
                ")",
                "((1 + 2) * 3",
                "(1 + 2) * 3)",
                "5 + (10)"
            };

            //Act
            var results = testStrings
                .Select(s => new { String = s, Result = ArithmeticParser.Parse(s)})
                .ToList();

            //Assert
            var output = new Output();
            var report = results.AsReport(rep => rep
                .AddColumn(r => r.String, cc => cc.Heading("Input"))
                .AddColumn(r => r.Result.IsValid, cc => { })
                .AddColumn(r => r.Result.Describe(), cc => cc.Heading("Result"))
                .AddColumn(r => r.Result.ErrorMessage, cc => {})
            );

            output.FormatTable(report);
            output.Report.Verify();
        }

        [Fact]
        public void ResultIsComputed()
        {
            //Arrange
            var testStrings = new [] {
                "5 + 3 * 6",
                "5 * 3 + 6",
                "5 * 2 + 3 / 4",
            };

            //Act
            var results = testStrings
                .Select(s => new { String = s, Result = ArithmeticParser.Parse(s)})
                .ToList();

            //Assert
            var output = new Output();
            output.FormatTable(results.Select(r => new { r.String,  Description = r.Result.Describe(), Result = r.Result.Compute()}));
            output.Report.Verify();
        }
    }
}