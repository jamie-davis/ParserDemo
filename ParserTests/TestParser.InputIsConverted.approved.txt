                  Is                                                                           
Input             Valid Result                  Error Message                                  
----------------- ----- ----------------------- -----------------------------------------------
1                 True  1                                                                      
32-5              True  -[32, 5]                                                               
(100 * 3)         True  (*[100, 3])                                                            
100 + (3 * 5)     True  +[100, (*[3, 5])]                                                      
(3 * 5) + 100     True  +[(*[3, 5]), 100]                                                      
1 + 2 + 3         True  +[1, +[2, 3]]                                                          
1 + (3 * (5 + 6)) True  +[1, (*[3, (+[5, 6])])]                                                
1 +               False                         Unable to interpret calculation at "+"         
(1 + 3            False                         Unable to interpret calculation at "( 1 + 3"   
(                 False                         Unable to interpret calculation at "("         
)                 False                         Unable to interpret calculation at ")"         
((1 + 2) * 3      False                         Unable to interpret calculation at "( ( 1 + 2 )
                                                * 3"                                           
(1 + 2) * 3)      False                         Unable to interpret calculation at ")"         
5 + (10)          True  +[5, (10)]                                                             
                                                                                               